src/flag_gems/utils/libentry.py

```diff
@@ -14,47 +16,110 @@ def __init__(
         while not isinstance(fn, triton.runtime.JITFunction):
             fn = fn.fn
         self.jit_function: triton.runtime.JITFunction = fn
-        self.kernel_arg_indices = []
+        self.spec_indices = []
+        self.dns_indices = []
         for p in self.jit_function.params:
             if not p.is_constexpr:
-                self.kernel_arg_indices.append(p.num)
+                if p.do_not_specialize:
+                    self.dns_indices.append(p.num)
+                else:
+                    self.spec_indices.append(p.num)
 
-    def run(self, *args, **kwargs):
-        key = []
-        for arg in args:
+    def key(self, spec_args, dns_args, const_args):
+        entry_key = []
+        for arg in spec_args:
+            if hasattr(arg, "data_ptr"):
+                entry_key.append(str(arg.dtype))
+                entry_key.append(arg.data_ptr() % self.divisibility == 0)
+            else:
+                entry_key.append(type(arg))
+                entry_key.append(arg)
```

Comment:
List comprehension is an immutable fashion for coding loops. It's normally more efficient than the mutable counterpart.

Reply:
done